### ISSUE TITLE
quantizer

### ISSUE THREAD

**Issue by jerch** [photopea/UPNG.js#53]
@photopea Thx for the great quantizer! It is really fast and shows very good results. Currently I am trying to use it for my SIXEL encoding package (this PR https://github.com/jerch/node-sixel/pull/16).

A few questions/remarks:
- Your exact color matching is quite slow (understandable, since it has to walk parts of the kd tree). I found populating the palette colors in 16x16x16 cubes in the colorspace and doing local brute force searching being almost as fast as your fast approximation, but with much better results (See `ColorMatcher` class). The search itself is spheric (euclidian distance) and is done on the outer cube sphere first, and for uncertain colors (those outside of an inner subsphere) with a second bigger sphere. The cut-down of the search space works nicely with good palette color distribution, but ofc falls short, if they all aggregate within a few cubes, or the palette is rather big (>1024 colors). For that one could introduce subcubes (again forming an octree), but I have not implemented this (not needed for SIXEL, as the palette should not be >256).
- The quantizer has issues to build a 512 color palette for this image: https://github.com/jerch/node-sixel/blob/master/palette.png, it always stops at 503 colors. Maybe you want to have a look at it.
- Is there a way to strip down the quantizer in terms of channel resolution and channels being used? SIXEL only allows RGB values in full percent limiting the color space to ~1M colors, also alpha is not supported (beside marking one palette entry as fully transparent). Thus I started to wonder, if the calculation can be simplified with those restrictions in mind. This might not work, as you calc most things in float internally, not sure if a partial rewrite with discrete integer calcs is worth the shot (JS also makes this really hard).
- As already mentioned in the other issue - could you update the npm package somehow? Directly pulling source code from repos is cumbersome :smile_cat:.

Edit: I also implemented a very small dithering with forward error diffusion, with does not penalize image size and runtime alot, but helps to remove most obvious bandings.

**Comment by photopea** [comment]
Hi, are you sure you are using the latest source code, which I uploaded to GitHub five days ago?

When finding the closest color for each palette pixel, there is a fast and a slow (exact) method. The exact is used when the input is smaller than 20 MB.

**Comment by jerch** [comment]
@photopea Seems so, I pulled the code at commit cb463cb7b857fc1fbd47bb8628c536e02aaa24ce. There is one later thought, which lists no changes.

> When finding the closest color for each palette pixel, there is a fast and a slow (exact) method. The exact is used when the input is smaller than 20 MB.

Yes I saw that branching, tbh - it looks weird to me, not really backed up by any metrics beside calc penalty for "image being big"?

**Comment by photopea** [comment]
Hi, there was a bug when computing the highest eigenvector of a matrix. It happened for "regular" inputs such as yours.

I think I fixed it. Could you test a new version?

**Comment by jerch** [comment]
Yes works. :+1:

**Comment by photopea** [comment]
And what about other problems? Could you check if they are fixed now?

**Comment by jerch** [comment]
There are not really issues left beside the npm package update.

About finding the nearest color:
I willl stick with my cubic partitions, as it is as fast as your fast approximation for images > 20 MB, but gives closer colors as in your exact search.

**Comment by photopea** [comment]
I am not a user of npm and I usually dont like to learn how to use new native programs. If you put it on NPM, I can provide a link in our README.

I don't know what you mean by cubic partitions. If the palette is given, UPNG.js should find the closest palette color to each pixel, so when you say you find even closer colors, it does not make any sense to me :/

**Comment by jerch** [comment]
> ... UPNG.js should find the closest palette color to each pixel ...

Your fast approximation does not imho, it cuts the search only to visited subtrees missing closer colors on neighboring paths.

**Comment by photopea** [comment]
@jerch Yes, that is true, our fast method does not find the closest color.

So your method is faster than our fast method and finds closer colors?

**Comment by jerch** [comment]
Yes, it is basically as fast as your fast one, but finds closer matches as the slower one.

**Comment by photopea** [comment]
Interesting. Could you describe your method? Are you sure it works well with all kinds of palettes (images)? What is the time complexity of finding a color?

**Comment by jerch** [comment]
@photopea Sure. It is basically spatial partitioning into 16x16x16 cubes of the RGB room (thus 4096 cubes). Every cube maps to 2 palette color "clouds" - first one containing all palette colors within the outer cube sphere or (nearest if empty), second one within the sphere of r=42 (~ the outer sphere of 4 cubes). The mappings are created once after palette creation.
For color matching the cube key is simply derived from the higher 4 color bits, on cube level I simply do a brute force euclidian distance test against the first sphere points, and if the distance of the tested color to the cube center is bigger than my certainty range (r=8), I test also against the second sphere colors.

So basically this is a cubic approximation of a spherical problem with 2 certainty levels. The error rate can be changed by changing the radius of both spheres. The values I ended up with are about as fast as your fast approx., but with error rates below 1% typically. I also tested other cube resolutions, but found the 16x16x16 cubes as sweep spot for well distributed palettes with 64 and 256 colors.

Runtime behavior:
- cube construction with color mappings: ~20 ms (thats fairly stable even for different palette sizes, seems the cube amount is the dominating factor (it is ~80 ms for  32768 cubes of 8x8x8 size, or <5ms for 512 cubes of 32x32x32)
- color matching:
  - cube key derivation in O(1)
  - brute force euclidian distance: O (n) for n colors in the inner sphere, for well distributed palette colors O(1) (cube maps only to 1 color)

The speed benefit comes from the fact, that the partitioning cuts off most palette colors to be tested, roughly by 4096 (not quite that much as the sphere covers more than the cube itself). Furthermore most cubes will only hold one color, which makes the match for color within those cubes O(1). And here is the catch - for lousy distributed palettes the algo will fall short, the worst case being all palette colors being populated into one cube - boom, again full ED testing. To avoid that, one can introduce subcubes, if the color count gets too high for a single cube. Note that this somewhat forms an octree, and yes I basically took the partitioning idea from my octree experiments. Still the certainty areas of the spherical search around the cubes is abit tricky.

**Comment by photopea** [comment]
What you describe is usually called a "Grid structure". It works well, if you are sure, that the samples are "uniformly distributed", without many "clusters" with a high density.

If your are sure that your palettes are always so nice, you can keep using it. However, I am afraid to use such method, as I think palettes are rarely that "nice" for classic images. In the worst case, you would have O(n) for every search.

If you decide to switch to octrees (for either exact or approximate search), I don't think it would be any faster than the methods that are currently in UPNG.js .

**Comment by jerch** [comment]
> If your are sure that your palettes are always so nice, you can keep using it.

I disagree, it is not a matter of nice palette color distribution, but how far one goes into subcubes if the higher level cube still holds too many colors. If you restrict "colors per cube" to only 4 you would have to go two bits deeper only. The benefit compared to octrees is the fact, that such a tree has a higher base than 2, thus can faster exclude uninteresting parts of the tree. The search is like O(log2 n) vs. O(logx n) for x >> 2. Note that the higher tree branching does not impact runtime at color matching phase, as the cube key can always be derived in O(1).
Ofc higher cube resolution impacts the construction phase - it is trading color matching runtime for initial runtime and space (but this happens only once). And the fact, that any cube always contains a nearest color cuts the expensive neighboring checks/search, which are always needed otherwise to find the real the closest match.

Note that for exact match your algo also has a worst case of O(n), if the tree happens to be quite unbalanced or the neighboring search ends up including all nodes (given you do proper ascending neighboring checks).

**Comment by photopea** [comment]
It is hard for me to argue with you, as I can not see your implementation.

Could you test your method on at least 30 different photographs? Is it always faster than our fast method? Or if it is equally fast, does it always find closer neighbours (i.e. the total error of the image is lower)?

**Comment by jerch** [comment]
Sorry, this is getting redundant and I dont know how to answer to that beside saying, that everything is already stated above (including a pointer to my simplified impl). Whether you want to try that (or a modified version with better skewed color distribution support) out for your needs is up to you, I am not in the position to try to "convince" you.

**Comment by photopea** [comment]
Ok, I still think, that our implementation works better than yours in practical cases. I will close it now.

But thanks for letting us know about the bug with finding a 512 item palette.

**Comment by jerch** [comment]
> Ok, I still think, that our implementation works better than yours in practical cases.
 
Yes please have it your way, although runtime and accuracy speaks otherwise.

**Comment by photopea** [comment]
You never mentioned on how many and what kind of images (palettes) you tested it on. I think you can not make such claims without testing it properly on many inputs.

**Comment by jerch** [comment]
A test with several pictures does not prove anything, just math/algo thoughts can. It has been shown that an octree with local spherical search can find the nearest in O(log log n), which is almost constant. Plus it being fixed, thus easy to derive the keys, which cuts down the work for a single step by a constant factor. Now do the maths - which a higher branching factor, the log gets a higher base. And so on...
I know that O(log1000 log1000 (n/1000)) is still closer to O(log n) than anything else, but those shifts in constants is what makes r-trees and other weird spatial tree concepts showing an advantage over more general concepts like a "dimension-rotating binary tree".

But if you insist - I tested this with ~50 images, from real pictures, greyscale with/without color tint (sepia) to hard to digest graphics with gradients. Btw the hardest one in terms of wrong colors was my palette image with ~12% error rate compared to a full NN search. For most pictures the error rate was far below 1%, not like your fast approx with 10 to 30% for my set of pictures.

**Comment by friksa** [comment]
Guys, clearly you are both smart.  Let's drop the ego and work together for the best solution.  No one wins in an argument.  Keep up the great work!

**Comment by photopea** [comment]
@jerch Do you think you could extend your grid method to make it find the closest color? That would be really useful.

**Comment by jerch** [comment]
@photopea Yes, I think that would be possible with true voronoi tessellation during construction, will see if thats still ok in terms of construction runtime.
